### ISSUE TITLE
hooks from one plugin overlap hooks from other plugins?

### ISSUE THREAD

**Issue by ambs** [PerlDancer/Dancer2#887]
When using 

`````
   use Dancer2::Plugin::Database;
   use Dancer2::Plugin::Feed;
```````
the hooks from the first module disappear.
If I change the order, they work.

When looking to the hooks, i found out that:

````
 -- checking hooks in Dancer2::Core::DSL__WITH__Dancer2::Plugin::Database__WITH__Dancer2::Plugin::Feed=HASH(0x7f8aded490b8) --
$VAR1 = {};
````
when Database is used first, and
```````
 -- checking hooks in Dancer2::Core::DSL__WITH__Dancer2::Plugin::Feed__WITH__Dancer2::Plugin::Database=HASH(0x7ff289f88c78) --
$VAR1 = {
          'plugin.database.database_connection_lost' => [],
          'plugin.database.database_connection_failed' => [],
          'plugin.database.database_connected' => [],
          'plugin.database.database_error' => []
        };
`````````
when used last.

Still no idea how to fix it, as I do not understand the plugin code. But it seems the hook list for plugins is reset on each plugin import.

**Comment by ambs** [comment]
Let me bring @bigpresh to the ticket, as he is an interested party on this.

**Comment by ambs** [comment]
More info:

````
use Dancer2;
use Dancer2::Plugin::Feed;
use Dancer2::Plugin::Emailesque;
use Dancer2::Plugin::Deferred;
use Dancer2::Plugin::Database;
`````

I get
`````
Hook 'plugin.database.database_connected' does not exist at /opt/perl/lib/site_perl/5.20.1/Dancer2/Plugin/Database.pm line 43.
`````

Changing the order of the last two:
```````
use Dancer2;
use Dancer2::Plugin::Feed;
use Dancer2::Plugin::Emailesque;
use Dancer2::Plugin::Database;
use Dancer2::Plugin::Deferred;
````````
I get
``````
 'database_connected' does not exist at /opt/perl/lib/site_perl/5.20.1/Dancer2/Plugin/Database.pm line 43.
``````

I think this change of name is not intentional and is, at the least, weird.

**Comment by veryrusty** [comment]
Also see #510 for some further detail about issues with plugin hooks.

@ambs this is part of why there is some effort to rework plugins. Have you have a chance to look at either proposal?

**Comment by ambs** [comment]
Not really. Unfortunately. At the moment I just needed an workaround.

**Comment by ambs** [comment]
More info on this.

Into `Dancer2::Core::Role::Hookable::_build_hooks` added  a dumper with the hooks being built for each plugin/engine. Also, in `Dancer2::Plugin` added a method to know where the `current_hooks` are registered.

So, the `_build_hooks` seems to be called before the plugin defining its hooks:

``````
building hooks on Dancer2::Core::DSL__WITH__Dancer2::Plugin::Feed__WITH__Dancer2::Plugin::Emailesque__WITH__Dancer2::Plugin::Deferred=HASH(0x7ff8b564ebc8)
$VAR1 = {};
Defning Dancer2::Plugin::Database::supported_hooks
`````````
So, the definition of the hooks gets called before the plugin gets loaded (correct), but they are no more built again.

Not sure if I made me clear. But if I could understand better, I could explain better :-)

**Comment by ambs** [comment]
`use Dancer2::Plugin::Deferred` and `use Dancer2::Plugin::Database` do not play together in any order. And the first does not define any hook.

**Comment by ambs** [comment]
I think I understood the problem. If `Deferred` is used first, and because it register code to be run in hooks, Moo creates `_build_hook` lazily. Then, when Database is loaded, new hooks are not considered as `_build_hooks` was already populated.

**Comment by ambs** [comment]
#889 solves the problem, it seems.

**Comment by hvoers** [comment]
On Sat, 18 Apr 2015, Russell Jenkins wrote:

> Have you have a chance to look at either proposal?

I searched for this but did not find the proposals.

Do you have any links?

**Comment by ambs** [comment]
I think they were discussed in the developers mailing list, only.

I would be grateful if this fix could be integrated into Dancer2 while a better plugin mechanism is not in place.

**Comment by ambs** [comment]
The bad news is that it doesn't work:
````````````
Error while loading /Users/ambs/MyReps/Dicionario-Aberto/www2/bin/app.psgi: Hook 'plugin.database.database_connected' does not exist at /opt/perl/lib/site_perl/5.20.1/Dancer2/Plugin/Database.pm line 43.
``````````````
I am not sure I will be able to look to the code right now, but in case you have the time before me, try to test with:
`````````
use Dancer2::Plugin::Deferred;
use Dancer2::Plugin::Database;
``````````
Cheers.
```````````

**Comment by xsawyerx** [comment]
I'm closing this for now, as it has been resolved by #892.
